datafusion/expr/src/type_coercion/functions.rs

```diff
@@ -50,17 +50,23 @@ pub fn data_types_with_scalar_udf(
     func: &ScalarUDF,
 ) -> Result<Vec<DataType>> {
     let signature = func.signature();
+    let type_signature = &signature.type_signature;
 
     if current_types.is_empty() {
-        if signature.type_signature.supports_zero_argument() {
+        if type_signature.supports_zero_argument() {
             return Ok(vec![]);
         } else {
-            return plan_err!("{} does not support zero arguments.", func.name());
+            if type_signature.used_to_support_zero_arguments() {
```

Comment:
Unless this check is present, the error messages about TypeSignature::nullary are confusing as they are shown when the function is called with zero args, even if its signature requires something else (like exact signatures, etc)